File: done/cpp/euler_004.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
using namespace std;

int reverse(int num)
{
    int digit = 0;
    int rev = 0;
    int n = num;
    while (n != 0)
    {
        digit = n % 10;
        rev = (rev * 10) + digit;
        n = n / 10;
    }
    return rev;
}

int p004()
{
    int max_pal = 0;
    for (int i = 100; i < 1000; i++)
    {
        for (int j = 100; j < 1000; j++)
        {
            int m = i * j;
            if (m == reverse(m) && m > max_pal)
            {
                max_pal = m;
            }
        }
    }
    return max_pal;
}

int main()
{
    cout << p004() << endl;
    return 0;
}

File: done/js/euler_004.js
const isPalindrome = require('../../utils/utils').isPalindrome;

const p004 = (min = 900) => {
  for (let i = 999; i > min; i--) {
    for (let j = 999; j > min; j--) {
      const num = i * j;
      if (isPalindrome(num)) {
        return num;
      }
    }
  }
  return -Infinity;
};

console.log(p004());
